File: demos/binarytrees/src/binarytrees.js
"use strict";

const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');

function TreeNode(left, right) {
    this.left = left;
    this.right = right;
}

function itemCheck(node) {
    if (node.left === null) return 1;
    return 1 + itemCheck(node.left) + itemCheck(node.right);
}

function bottomUpTree(depth) {
    return depth > 0
        ? new TreeNode(bottomUpTree(depth - 1), bottomUpTree(depth - 1))
        : new TreeNode(null, null);
}

if (isMainThread) {
    const maxDepth = Math.max(6, parseInt(process.argv[2]) || 0);
    const stretchDepth = maxDepth + 1;

    const check = itemCheck(bottomUpTree(stretchDepth));
    console.log(`stretch tree of depth ${stretchDepth}\t check: ${check}`);

    const longLivedTree = bottomUpTree(maxDepth);

    const results = [];
    let tasksRemaining = 0;

    for (let depth = 4; depth <= maxDepth; depth += 2) {
        const iterations = 1 << (maxDepth - depth + 4);
        tasksRemaining++;
        
        // Use worker threads for the parallelizable part of the benchmark
        const worker = new Worker(__filename, {
            workerData: { iterations, depth }
        });

        worker.on('message', (msg) => {
            results[depth] = `${iterations}\t trees of depth ${depth}\t check: ${msg}`;
            tasksRemaining--;
            if (tasksRemaining === 0) {
                for (let d = 4; d <= maxDepth; d += 2) {
                    console.log(results[d]);
                }
                console.log(`long lived tree of depth ${maxDepth}\t check: ${itemCheck(longLivedTree)}`);
            }
        });
    }
} else {
    // Worker thread logic
    const { iterations, depth } = workerData;
    let check = 0;
    for (let i = 0; i < iterations; i++) {
        check += itemCheck(bottomUpTree(depth));
    }
    parentPort.postMessage(check);
}

File: demos/binarytrees/src/binarytrees.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// contributed by 1011X
// contributed by Michal Muskala
// modified by Jiri Vymazal
// modified by Jeremy Zerfas

extern crate rayon;
extern crate typed_arena;

use rayon::prelude::*;
use typed_arena::Arena;

struct Tree<'a> {
    left: Option<&'a Tree<'a>>,
    right: Option<&'a Tree<'a>>,
}

fn item_check(tree: &Tree) -> i32 {
    if let (Some(left), Some(right)) = (tree.left, tree.right) {
        1 + item_check(left) + item_check(right)
    } else {
        1
    }
}

fn bottom_up_tree<'a>(arena: &'a Arena<Tree<'a>>, depth: i32) -> &'a Tree<'a> {
    let tree = arena.alloc(Tree {
        left: None,
        right: None,
    });
    if depth > 0 {
        tree.left = Some(bottom_up_tree(arena, depth - 1));
        tree.right = Some(bottom_up_tree(arena, depth - 1));
    }
    tree
}

fn main() {
    let n = std::env::args()
        .nth(1)
        .and_then(|n| n.parse().ok())
        .unwrap_or(10);
    let min_depth = 4;
    let max_depth = if min_depth + 2 > n { min_depth + 2 } else { n };
    let stretch_depth = max_depth + 1;

    {
        let arena = Arena::new();
        let stretch_tree = bottom_up_tree(&arena, stretch_depth);
        println!(
            "stretch tree of depth {}\t check: {}",
            stretch_depth,
            item_check(stretch_tree)
        );
    }

    let long_lived_arena = Arena::new();
    let long_lived_tree = bottom_up_tree(&long_lived_arena, max_depth);

    let messages: Vec<_> = (min_depth..=max_depth)
        .into_par_iter()
        .step_by(2)
        .map(|depth| {
            let iterations = 1 << (max_depth - depth + min_depth);
            let mut check = 0;

            for _ in 0..iterations {
                let arena = Arena::new();
                let tree = bottom_up_tree(&arena, depth);
                check += item_check(tree);
            }

            format!("{}\t trees of depth {}\t check: {}", iterations, depth, check)
        })
        .collect();

    for message in messages {
        println!("{}", message);
    }

    println!(
        "long lived tree of depth {}\t check: {}",
        max_depth,
        item_check(long_lived_tree)
    );
}
